posts/2019-01-14-definitive-guide-terraform-serverless.md

```diff
@@ -0,0 +1,209 @@
+
+To which we say: you’re absolutely right. Many companies using Serverless already use Terraform, and some Serverless Framework functionality, especially when it comes to provisioning cloud resources, is similar to what Terraform can do.
+
+So if both Terraform and Serverless can solve your infrastructure automation needs, which one should you use? And should you use just one for all purposes?
+
+We’ve got answers.
+
+In this article, we’ll talk about the right way to manage infrastructure when using both Terraform and Serverless, and check out a real-world example of integrating Terraform and Serverless in a project.
+
+#### Why automate infrastructure management
+
+Infrastructure as Code (IaC) becomes really important once developers need a way to organize their growing cloud infrastructure and collaborate across teams.
+
+Most importantly, IaC tools make it necessary to have process and discipline; there’s a smaller chance of accidental or unexpected changes, and it’s easier to share configuration between different parts of your infrastructure.
+
+#### Managing shared vs. app-specific infrastructure
+
+While we believe that all infrastructure should be managed with IaC automation, we like to distinguish between the infrastructure that’s specific to one application and the infrastructure that’s shared between multiple applications in your stack. Those might need to be managed in different ways.
+
+Application-specific infrastructure gets created and torn down as the app gets deployed. You rarely change a piece of application-specific infrastructure; you’ll just tear everything down and re-create it from scratch. As the app is developed, the infrastructure that supports it also needs to change, sometimes significantly from one deploy to another.
+
+The shared infrastructure, on the other end, rarely gets re-created from scratch and is more stateful. The core set of infrastructure (such as the set of security groups and your VPC ID), won’t change between the deploys of your application, as they’re probably referenced by many applications in your stack. Those more persistent pieces of infrastructure will generally be managed outside of your deploy pipeline.
+
+So, application-specific and shared infrastructure are different enough that they should be managed with different tools.
+
+#### Serverless vs Terraform: when to use which
+
+For an organization using both Terraform and Serverless, here are the benefits of each, and when you should choose one over the other.
+
+##### Serverless for app-specific infrastructure
+
+For application-specific infrastructure, we suggest managing all the pieces with the Serverless Framework, for a few reasons.
+
+First, you couple this infrastructure to the application itself. Second, we like to think that the application "owns" things, like the tables in the Postgres database. There is little value in managing the table names outside of the application context (e.g. in Terraform).
```

Comment:
Could we add a `third` about application iteration?

`Third, you can iterate your application release without touching your shared infrastructure; software releases decouple from shared infrastructure allowing you to focus on the application itself without having to worry about infrastructure changes`